File: go-manychain/core/address.go
package core

import (
	"crypto/sha256"
	"manychain/crypto"
)

// TODO: address can be more complex then this I think. It can have smart data, it can be a signed object
// TODO: can everything be an object with an address?

type Address struct {
	value [20]byte
}

func (a Address) Value() [20]byte {
	return a.value
}

func New() Address {
	ds := crypto.NewDigitalSignatureKeys()
	address := FromKey(ds.PublicKey())
	return address
}

func FromKey(publicKey crypto.PublicKey) Address {
	bytes := publicKey.Bytes()
	h := sha256.Sum256(bytes)
	var v [20]byte
	for i := 0; i < 20; i++ {
		v[i] = h[i]
	}

	return Address{
		value: v,
	}
}

File: zigtut/src/core/address.zig
const std = @import("std");
const cr = std.crypto;

pub const Address = struct {
    value : *const [65]u8,

    pub fn new() Address {
        const kp = cr.sign.ecdsa.EcdsaP256Sha256
        .KeyPair.generate();

        const bytes   = kp.public_key.toUncompressedSec1();

        return Address {
            .value = &bytes
        };
    }

    pub fn Value(self: Address) *const [65]u8   {
        return self.value;
    }
};